src/mpid/ch4/shm/xpmem/xpmem_pre.h

```diff
@@ -14,10 +14,32 @@
 #include <xpmem.h>
 
 #define MPIDI_XPMEM_PERMIT_VALUE ((void *)0600)
+#define MPIDI_XPMEM_SEG_PREALLOC 8      /* Number of segments to preallocate in the "direct" block */
+
+typedef struct MPIDI_XPMEM_seg {
+    /* AVL-tree internal components start */
+    struct MPIDI_XPMEM_seg *parent;
+    struct MPIDI_XPMEM_seg *left;
+    struct MPIDI_XPMEM_seg *right;
+    uint64_t height;            /* height of this subtree */
+    /* AVL-tree internal components end */
```

Comment:
This AVL tree logic is made specifically for XPMEM segment (such as determining segment overlap cases, segment insert, segment delete and segment reuse), that's why it is part of XPMEM. If other code need AVL tree too, it should use its own logic to implement its own AVL tree (search, insert, delete operations).

Reply:
I see, yes now that I look at it better it has very specific implementation. In principle it would be possible to decouple the two implementations but this might result in some performance degradation.